tests/test_invenio_ldapclient.py

```diff
@@ -127,38 +131,35 @@ def test_view_ldap_conn_returns_True(app):
     app.extensions['security'] = Mock()
     InvenioLDAPClient(app)
     app.config['SECURITY_POST_LOGIN_VIEW'] = '/abc'
-
+    app.config['WTF_CSRF_ENABLED'] = False
     ldap_conn = Mock(bind=lambda: True, unbind=lambda: True)
     user = Mock()
-    with patch(
-        'invenio_ldapclient.views._ldap_connection',
-        autospec=True, return_value=ldap_conn
-    ) as ldap_conn_mock:
-        with patch(
-            'invenio_ldapclient.views._find_or_register_user',
-            autospec=True, return_value=user
-        ) as find_register_mock:
-            with patch(
-                'invenio_ldapclient.views.after_this_request',
-                autospec=True
-            ) as after_request_mock:
-                res = app.test_client().post(
-                    "/ldap-login",
-                    data=dict(username='itsame', password='good')
-                )
-    lform = ldap_conn_mock.call_args[0][0]
-    assert lform.username.data == 'itsame'
-    assert lform.password.data == 'good'
-    assert ldap_conn_mock.called is True
-    find_register_mock.assert_called_once_with(ldap_conn, 'itsame')
-    after_request_mock.assert_called_once_with(
-        invenio_ldapclient.views._commit
-    )
 
-    assert app.view_functions['security.login'] == \
-        invenio_ldapclient.views.ldap_login_form
-    assert res.status_code == 302
-    assert res.location == 'http://localhost/abc'
+    @patch('invenio_ldapclient.views.after_this_request', autospec=True)
+    @patch('invenio_ldapclient.views._find_or_register_user', autospec=True,
+           return_value=user)
+    @patch('invenio_ldapclient.views._ldap_connection', autospec=True,
+           return_value=ldap_conn)
+    def patched_test(ldap_conn_mock, find_register_mock, after_request_mock):
+        res = app.test_client().post(
+            "/ldap-login",
+            data=dict(username='itsame', password='good')
+        )
+
+        lform = ldap_conn_mock.call_args[0][0]
+        assert lform.username.data == 'itsame'
+        assert lform.password.data == 'good'
+        assert ldap_conn_mock.called is True
+        find_register_mock.assert_called_once_with(ldap_conn, 'itsame')
+        after_request_mock.assert_called_once_with(
+            invenio_ldapclient.views._commit
+        )
+        assert app.view_functions['security.login'] == \
+            invenio_ldapclient.views.ldap_login
+        assert res.status_code == 302
+        assert res.location == 'http://localhost/abc'
+
+    patched_test()
```

Comment:
Is this how python testing works? You need to call `patched_test()` here, but `test_view_ldap_conn_returns_True()` is automatically called?

Reply:
Yes. Top level functions (or Classes) starting with `test_` are automatically called by the test runner. If you have nested functions inside those functions, you need to call them yourself. And the reason why we have a nested function in this case is to take advantage of explicit mocking (by doing assertions on the Mock objects after the fact).

Here I just refactored code to be more flat rather than nested. It is more readable (more "pythonic": Flat is better than nested. See `python -m this`) than the gnarly nesting.

Reply:
Got it, thanks.